### PR TITLE
Locked account

### PR DESCRIPTION
**Pull request by shane-tomlinson**
Continue with @chilts #91, now based on the "reverse" refactor.

Goes along with:
* https://github.com/mozilla/fxa-auth-db-mem/pull/11
* https://github.com/mozilla/fxa-auth-db-mysql/pull/7
* https://github.com/mozilla/fxa-auth-mailer/pull/23
* https://github.com/mozilla/fxa-js-client/pull/141
* https://github.com/mozilla/fxa-content-server/pull/1806
* https://github.com/mozilla/fxa-auth-server/pull/867

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
@chilts and @rfk - this one could use a review as well. :cake:

File: test/backend/db_tests.js [hunk 1]
```diff
@@ -450,6 +451,49 @@ module.exports = function(config, DB) {
...
+                return db.unlockCode(uid)
+              })
+              .then(function (_unlockCode) {
+                t.deepEqual(_unlockCode.unlockCode, unlockCode, 'unlockCode should be set')
+
+                // try to unlock the account
+                return db.unlockAccount(uid)
+              })
+              .then(null, function(err) {
+                t.fail('We should not have failed this .unlockAccount() request' + String(err));
```
Comment (chilts) [inline test/backend/db_tests.js]: I never know what the right answer is here and I struggled with this myself in a previous issue related to deleting something. If you "unlock something that is already unlocked, then yeah, we're already done". ie. in my previous case, if I delete something which isn't there, we're done and we're succesful. Am just thinking out loud and not suggesting anything other than what you already have - just wondering what the best thing is here. An idempotent call would be nice here, but being explicit is also good. End of ramblings - feel free to ignore. :D

File: test/backend/db_tests.js [hunk 2]
```diff
@@ -450,6 +451,49 @@ module.exports = function(config, DB) {
...
+              })
+              .then(function(result) {
+                t.deepEqual(result, {}, 'Returned an empty object for unlockAccount')
+                // now check it's been saved
+                return db.emailRecord(email)
```
Comment (chilts) [inline test/backend/db_tests.js]: In many other tests where the accounts table has changed, I try and then test both `.account()` and also `.emailRecord()` at the same time to make sure they return what we expect. Therefore here it'd be nice if there was an `.emailRecord()` call above (to make sure the `lockedAt` is set and a `.account()` call here to also check it is now unset. Does that make sense? It's super specific but I wonder if it'll be more complete.

File: test/backend/db_tests.js [hunk 3]
```diff
@@ -589,6 +634,20 @@ module.exports = function(config, DB) {
...
+                return db.forgotPasswordVerified(PASSWORD_FORGOT_TOKEN_ID, ACCOUNT_RESET_TOKEN)
+              })
+              .then(function() {
+                t.pass('.forgotPasswordVerified() did not error')
+                // now check that the forgotPasswordVerified also reset the lockedAt
+                return db.emailRecord(Buffer(ACCOUNT.email))
+              })
+              .then(function(account) {
+                t.equal(account.lockedAt, null, 'account should now be unlocked')
                 t.pass('.createPasswordForgotToken() did not error')
```
Comment (chilts) [inline test/backend/db_tests.js]: Should this `t.pass()` now be in the function 3 stanzas up?

File: test/backend/db_tests.js [hunk 4]
```diff
@@ -695,11 +758,20 @@ module.exports = function(config, DB) {
...
-                return db.resetAccount(ACCOUNT.uid, ACCOUNT)
+                // lock the account to ensure the unlockCode is deleted
+                return db.lockAccount(uid, { lockedAt: lockedAt, unlockCode: unlockCode })
+              })
+              .then(function() {
+                t.pass('.lockAccount() did not error')
+                return db.unlockCode(uid)
+              })
+              .then(function(unlockCode) {
+                t.ok(unlockCode.unlockCode)
```
Comment (chilts) [inline test/backend/db_tests.js]: Nit: add a comment string so we know what the test is here (not just checking against null, but checking the unlockCode). I know you hate me. :-P

File: test/fake.js [hunk 5]
```diff
@@ -79,6 +79,8 @@ module.exports.newUserDataHex = function() {
     createdAt: Date.now()
   }
 
+  data.unlockCode = hex16()
```
Comment (chilts) [inline test/fake.js]: Can you add an equivalent line to the `newUserDataBuffer()` function too? This is only used in the prune_tokens test but as a matter of completeness.

Note: I'm totally to blame here for the duplicate functions and I'll look into having just one function generate the data and convert it into a Buffer/Hex equivalent. Sorry.

**Comment by chilts** [comment]
r+ here. Again, looks very nice. Nicely done @shane-tomlinson.

Testing locally passed 100% with the mem backend but I got one error on the mysql backend - it looks unrelated and I'll dig through it but I think it's my problem since it's not related to any of this.

File: test/backend/db_tests.js (hunk 3 above)
Comment (shane-tomlinson) [inline test/backend/db_tests.js]: yes, yes it should

File: test/backend/db_tests.js (hunk 4 above)
Comment (shane-tomlinson) [inline test/backend/db_tests.js]: >  I know you hate me. :-P

@chilts - not at all, this is what review is for!

File: test/backend/db_tests.js (hunk 1 above)
Comment (shane-tomlinson) [inline test/backend/db_tests.js]: I was not sure of the right way to go here either, I don't mind dropping errors on the ground as long as it's consistent with how the other DB functions operate. The one advantage to throwing an error here is we can show the user an "expired or invalid link" error if the link is no good w/o adding an extra backend call to check the validity of the unlockCode.

**Comment by shane-tomlinson** [comment]
@chilts - updated! Again, thanks for the critical eye and suggestions. I'm learning a lot from the guidance.

File: test/backend/db_tests.js (hunk 1 above)
Comment (chilts) [inline test/backend/db_tests.js]: I think you're right about being more explicit and returning the error to the client for the purposes of error reporting to the user. Let's leave this as you have it already.
